File: python/src/threads/loop/capabilities.py
"""The capability pre-check (spec/schema/README.md, "Capability pre-check"): a rendered request
the model can't take never becomes a `model_request`."""

from typing import Literal

from pydantic import JsonValue, TypeAdapter

from threads.loop.model import ModelInfo

type Unsupported = Literal["content_unsupported", "continuation_unsupported"]

_LINE: TypeAdapter[dict[str, JsonValue]] = TypeAdapter(dict[str, JsonValue])
_MEDIA = ("image_ref", "document_ref", "audio_ref")
_CONTINUATION = ("reasoning", "hosted_tool")


def mismatch(body: bytes, info: ModelInfo) -> Unsupported | None:
    """The first part of the request, in order, that the model doesn't declare it can take.
    The body is Render v1 the loop just rendered, so it parses."""
    for text in body.splitlines()[1:]:
        line = _LINE.validate_json(text)
        parts = line.get("content")
        for part in parts if isinstance(parts, list) else ():
            if not isinstance(part, dict):
                continue
            kind = part.get("type")
            media = line.get("role") in ("user", "tool") and kind in _MEDIA
            if media and kind not in info.accepts:
                return "content_unsupported"
            if kind in _CONTINUATION and part.get("provider") != info.model.provider:
                return "continuation_unsupported"
    return None

File: typescript/packages/core/src/model/capabilities.ts
import type { ModelInfo } from "./protocol";
import type { RenderRequest } from "./render-lines";

// a request whose parts the model can't take fails before dispatch
// with a typed code, so no attempt is recorded and nothing reaches the provider.

export type Unsupported = {
  readonly code:
    | "content_unsupported"
    | "continuation_unsupported"
    | "transport_fence_unsupported";
  readonly message: string;
};

const MEDIA = new Set(["image_ref", "document_ref", "audio_ref"]);

/** The first part of `request` that `info` doesn't declare it can take, if any. */
export function unsupported(
  request: RenderRequest,
  info: ModelInfo,
): Unsupported | undefined {
  const accepts = new Set<string>(info.accepts);
  const provider = info.model.provider;
  for (const line of request.lines) {
    if (line.role === "tools") continue;
    for (const part of line.content) {
      if (
        line.role !== "assistant" &&
        MEDIA.has(part.type) &&
        !accepts.has(part.type)
      )
        return {
          code: "content_unsupported",
          message: `${provider}/${info.model.name} does not accept ${part.type}`,
        };
      if (
        (part.type === "reasoning" || part.type === "hosted_tool") &&
        part.provider !== provider
      )
        return {
          code: "continuation_unsupported",
          message: `a ${part.provider} ${part.format} part can't be sent to ${provider}`,
        };
    }
  }
  return undefined;
}
